pkg/codec/vaapi/vp8.go

```diff
@@ -0,0 +1,494 @@
+package vaapi
+
+// reference: github.com/intel/libva-utils/encode/vp8enc.c
+
+// #cgo pkg-config: libva libva-drm
+// #include <fcntl.h>
+// #include <stdint.h>
+// #include <stdio.h>
+// #include <stdlib.h>
+// #include <string.h>
+// #include <unistd.h>
+//
+// #include <va/va.h>
+// #include <va/va_drm.h>
+// #include <va/va_enc_vp8.h>
+//
+// #include "helper.h"
+//
+// void setShowFrameFlagVP8(VAEncPictureParameterBufferVP8 *p, uint32_t f) {
+//   p->pic_flags.bits.show_frame = f;
+// }
+// void setForceKFFlagVP8(VAEncPictureParameterBufferVP8 *p, uint32_t f) {
+//   p->ref_flags.bits.force_kf = f;
+// }
+// void setFrameTypeFlagVP8(VAEncPictureParameterBufferVP8 *p, uint32_t f) {
+//   p->pic_flags.bits.frame_type = f;
+// }
+// void setRefreshLastFlagVP8(VAEncPictureParameterBufferVP8 *p, uint32_t f) {
+//   p->pic_flags.bits.refresh_last = f;
+// }
+// void setCopyBufferToGoldenFlagVP8(VAEncPictureParameterBufferVP8 *p, uint32_t f) {
+//   p->pic_flags.bits.copy_buffer_to_golden = f;
+// }
+// void setCopyBufferToAlternateFlagVP8(VAEncPictureParameterBufferVP8 *p, uint32_t f) {
+//   p->pic_flags.bits.copy_buffer_to_alternate = f;
+// }
+import "C"
+
+import (
+	"errors"
+	"fmt"
+	"image"
+	"io"
+	"sync"
+	"unsafe"
+
+	"github.com/pion/mediadevices/pkg/codec"
+	mio "github.com/pion/mediadevices/pkg/io"
+	"github.com/pion/mediadevices/pkg/io/video"
+	"github.com/pion/mediadevices/pkg/prop"
+
+	"github.com/pion/webrtc/v2"
+)
+
+const (
+	surfaceVP8Ref0 = iota
+	surfaceVP8Ref1
+	surfaceVP8Ref2
+	surfaceVP8Ref3
+	surfaceVP8Input
+	surfaceVP8Num
+)
+
+type encoderVP8 struct {
+	r     video.Reader
+	buf   []byte
+	frame []byte
+
+	fdDRI    C.int
+	display  C.VADisplay
+	confID   C.VAConfigID
+	surfs    [surfaceVP8Num]C.VASurfaceID
+	ctxID    C.VAContextID
+	seqParam C.VAEncSequenceParameterBufferVP8
+	picParam C.VAEncPictureParameterBufferVP8
+	qMat     C.VAQMatrixBufferVP8
+	hrdParam hrdParam
+	frParam  frParam
+	rcParam  rcParam
+
+	frameCnt int
+	prop     prop.Media
+
+	mu     sync.Mutex
+	closed bool
+}
+
+func init() {
+	codec.Register(webrtc.VP8, codec.VideoEncoderBuilder(NewVP8Encoder))
+}
+
+// NewVP8Encoder creates new VP8 encoder
+func NewVP8Encoder(r video.Reader, p prop.Media) (io.ReadCloser, error) {
+	if (p.Width/16)*16 != p.Width || p.Width == 0 {
+		return nil, errors.New("width must be 16*n")
+	}
+	if (p.Height/16)*16 != p.Height || p.Height == 0 {
+		return nil, errors.New("height must be 16*n")
+	}
```

Comment:
Why are we dividing Width/Height by 16 and multiply it by 16 again?

Reply:
It's equivalent of 
```go
if p.Width%16 != 0 || p.Width == 0 {
```
this one would be better at readability. I'll update it.

This condition is due to that VP8/9 hardware accelerator only supports size multiple of 16.